### PR TITLE
Test Syncing with Another Fully Synced Service

### PR DESCRIPTION
**Pull request by nisdas**
This is part of #1593, 

- [x] Test Syncing with  another fully synced service.
- [x] Use Simulated Backend to generate blocks to initialize a service.
- [x] Use a Simulated P2P to send messages between the services.

### CODE REVIEW

File: beacon-chain/chaintest/backend/BUILD.bazel [hunk 1]
```diff
@@ -5,13 +5,12 @@ go_library(
...
         "fork_choice_test_format.go",
         "helpers.go",
-        "setup_db.go",
         "shuffle_test_format.go",
         "simulated_backend.go",
         "state_test_format.go",
     ],
     importpath = "github.com/prysmaticlabs/prysm/beacon-chain/chaintest/backend",
-    visibility = ["//beacon-chain/chaintest:__subpackages__"],
+    visibility = ["//beacon-chain:__subpackages__"],
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/BUILD.bazel]: Why change this?

Comment (prestonvanloon) [reply beacon-chain/chaintest/backend/BUILD.bazel]: Nevermind, i see now

File: beacon-chain/sync/querier.go [hunk 2]
```diff
@@ -24,8 +23,8 @@ type QuerierConfig struct {
 	ResponseBufferSize int
 	P2P                p2pAPI
 	BeaconDB           *db.BeaconDB
-	CurentHeadSlot     uint64
 	PowChain           powChainService
+	CurentHeadSlot     uint64
```
Comment (prestonvanloon) [inline beacon-chain/sync/querier.go]: Should be `CurrentHeadSlot`, typo

**Comment by prestonvanloon** [comment]
Overall looks good, a few minor comments and it seems that tests are failing.

File: beacon-chain/chaintest/backend/simulated_backend_test.go [hunk 3]
```diff
@@ -16,22 +17,22 @@ func TestSimulatedBackendStop_ShutsDown(t *testing.T) {
...
 
 func TestGenerateBlockAndAdvanceChain_IncreasesSlot(t *testing.T) {
 	backend, err := NewSimulatedBackend()
 	if err != nil {
-		t.Fatalf("Could not create a new simulated backedn %v", err)
+		t.Fatalf("Could not create a new simulated backend %v", err)
 	}
 
-	initialDeposits, privKeys, err := generateInitialSimulatedDeposits(params.BeaconConfig().SlotsPerEpoch)
+	privKeys, err := backend.SetupBackend(1000)
```
Comment (rauljordan) [inline beacon-chain/chaintest/backend/simulated_backend_test.go]: Can we run with fewer?

File: beacon-chain/sync/simulated_sync_test.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		simP2P.Send(&pb.ChainHeadResponse{
+			Slot: params.BeaconConfig().GenesisSlot + 10,
+			Hash: []byte{'t', 'e', 's', 't'},
+		}, p2p.Peer{})
+	}
+
+	return ss, beacondb
+}
+
+func TestSyncFromAFullySyncedNode(t *testing.T) {
```
Comment (rauljordan) [inline beacon-chain/sync/simulated_sync_test.go]: Follow the underscore pattern for test names here

File: beacon-chain/sync/regular_sync.go [hunk 5]
```diff
@@ -425,18 +424,18 @@ func (rs *RegularSync) handleBatchedBlockRequest(msg p2p.Message) {
 	response := make([]*pb.BeaconBlock, 0, endSlot-startSlot)
 
 	for i := startSlot; i <= endSlot; i++ {
-		block, err := rs.db.BlockBySlot(i)
+		retblock, err := rs.db.BlockBySlot(i)
```
Comment (terenc3t) [inline beacon-chain/sync/regular_sync.go]:
```suggestion
		retBlock, err := rs.db.BlockBySlot(i)
```

File: beacon-chain/sync/regular_sync.go [hunk 6]
```diff
@@ -425,18 +424,18 @@ func (rs *RegularSync) handleBatchedBlockRequest(msg p2p.Message) {
...
 		if err != nil {
 			log.Errorf("Unable to retrieve block from db %v", err)
 			continue
 		}
 
-		if block == nil {
+		if retblock == nil {
```
Comment (terenc3t) [inline beacon-chain/sync/regular_sync.go]:
```suggestion
		if retBlock == nil {
```

File: beacon-chain/sync/regular_sync.go [hunk 7]
```diff
@@ -425,18 +424,18 @@ func (rs *RegularSync) handleBatchedBlockRequest(msg p2p.Message) {
...
 			log.Debug("Block does not exist in db")
 			continue
 		}
 
-		response = append(response, block)
+		response = append(response, retblock)
```
Comment (terenc3t) [inline beacon-chain/sync/regular_sync.go]:
```suggestion
		response = append(response, retBlock)
```

File: beacon-chain/sync/simulated_sync_test.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+	numOfBlocks := 10
+	newP2P := &simulatedP2P{
+		subsChannels: make(map[reflect.Type]*event.Feed),
+		mutex:        new(sync.RWMutex),
+		ctx:          context.Background(),
+	}
+
+	// Sets up a synced service which has its head at the current
+	// numOfBlocks from genesis. The blocks are generated through
+	// simulated backend
```
Comment (terenc3t) [inline beacon-chain/sync/simulated_sync_test.go]:
```suggestion
	// simulated backend.
```

File: beacon-chain/sync/simulated_sync_test.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+	sub2 := us2.InitialSync.SyncedFeed().Subscribe(syncedChan2)
+	defer sub2.Unsubscribe()
+
+	highestSlot := <-syncedChan
+
+	highestSlot2 := <-syncedChan2
+
+	if highestSlot != uint64(numOfBlocks)+params.BeaconConfig().GenesisSlot {
+		t.Errorf("Sync services didnt sync to expectecd slot, expected %d but got %d",
```
Comment (terenc3t) [inline beacon-chain/sync/simulated_sync_test.go]:
```suggestion
		t.Errorf("Sync services didn't sync to expected slot, expected %d but got %d",
```

File: beacon-chain/sync/simulated_sync_test.go [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+			uint64(numOfBlocks)+params.BeaconConfig().GenesisSlot, highestSlot)
+	}
+
+	if highestSlot2 != uint64(numOfBlocks)+params.BeaconConfig().GenesisSlot {
+		t.Errorf("Sync services didnt sync to expectecd slot, expected %d but got %d",
```
Comment (terenc3t) [inline beacon-chain/sync/simulated_sync_test.go]:
```suggestion
		t.Errorf("Sync services didn't sync to expected slot, expected %d but got %d",
```

File: beacon-chain/sync/querier.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/sync/querier.go]: Alright, Done

File: beacon-chain/chaintest/backend/simulated_backend_test.go (hunk 3 above)
Comment (nisdas) [reply beacon-chain/chaintest/backend/simulated_backend_test.go]: Done

File: beacon-chain/sync/simulated_sync_test.go (hunk 4 above)
Comment (nisdas) [reply beacon-chain/sync/simulated_sync_test.go]: Done

File: beacon-chain/sync/regular_sync.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/sync/regular_sync.go]: Done for all of them

**Pull request merged**